[Clang][HIP] Target-dependent overload resolution in declarators and specifiers

ritter-x2a opened the pull request:
So far, the resolution of host/device overloads for functions in HIP/CUDA operates as if in a host-device context for code outside of function bodies, e.g., in expressions that are part of template arguments in top-level declarations. This means that, if separate host and device overloads are declared, the device overload is used in the device compilation phase and the host overload is used in the host compilation phase.

This patch changes overload resolution in such cases to prefer overloads that match the target of the declaration in which they occur. For example:

```cuda
__device__ constexpr int get_n() { return 64; }
__host__ constexpr int get_n() { return -1; }
__device__ std::enable_if<(get_n() > 32)>::type foo() { }
```

Before, this code would not compile, because `get_n` resolved to the host overload during host compilation, causing an error. With this patch, the call to `get_n` in the declaration of the device function `foo` resolves to the device overload in host and device compilation.

If attributes that affect the declaration's target occur after a call with target-dependent overload resolution, a warning is issued. This is realized by registering the Kinds of relevant attributes in the `CUDATargetContext` when they are parsed.

This is an alternative to PR #93546, which is required for PR #91478.

ritter-x2a committed:
[Clang][HIP] Target-dependent overload resolution in declarators and specifiers

So far, the resolution of host/device overloads for functions in HIP/CUDA
operates as if in a host-device context for code outside of function bodies,
e.g., in expressions that are part of template arguments in top-level
declarations. This means that, if separate host and device overloads are
declared, the device overload is used in the device compilation phase and the
host overload is used in the host compilation phase.

This patch changes overload resolution in such cases to prefer overloads that
match the target of the declaration in which they occur. For example:

__device__ constexpr int get_n() { return 64; }
__host__ constexpr int get_n() { return -1; }
__device__ std::enable_if<(get_n() > 32)>::type foo() { }

Before, this code would not compile, because get_n resolved to the host
overload during host compilation, causing an error. With this patch, the call
to get_n in the declaration of the device function foo resolves to the device
overload in host and device compilation.

If attributes that affect the declaration's target occur after a call with
target-dependent overload resolution, a warning is issued. This is realized by
registering the Kinds of relevant attributes in the CUDATargetContext when they
are parsed.

This is an alternative to PR #93546, which is required for PR #91478.

ritter-x2a committed:
fixup! [Clang][HIP] Target-dependent overload resolution in declarators and specifiers

Handle and test template functions outside of classes.

ritter-x2a committed:
fixup! fixup! [Clang][HIP] Target-dependent overload resolution in declarators and specifiers

Add a test to document the behavior for default arguments of template parameters.

Artem-B commented on clang/lib/Sema/SemaCUDA.cpp:
```diff
@@ -115,20 +143,65 @@ static bool hasAttr(const Decl *D, bool IgnoreImplicitAttr) {
          });
 }
 
+SemaCUDA::CUDATargetContext::CUDATargetContext(SemaCUDA *S,
```
This could probably be moved into the header.

ritter-x2a replied:
Addressed in e273a996854cb3256cfeed832c3f16ff6fd0f534.

Artem-B reviewed (commented):
Considering that we're adding another interesting quirk to how we interpret target attributes & function calls, it would be useful to run this by a language lawyer to make sure we're not missing something.

@zygoloid - would you have time to take a look or suggest who else may be the right person to consult on this kind of changes?

ritter-x2a committed:
fixup! fixup! fixup! [Clang][HIP] Target-dependent overload resolution in declarators and specifiers

Rename 'target specifier' -> 'target attribute', move CUDATargetContext constructor in header.

ritter-x2a commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -9017,6 +9017,10 @@ def err_global_call_not_config : Error<
   "reference to %select{__device__|__global__|__host__|__host__ __device__}0 "
   "%select{function|variable}1 %2 in %select{__device__|__global__|__host__|__host__ __device__}3 function">;
+def warn_target_specfier_ignored : Warning<
+  "target specifier has been ignored for overload resolution; "
```
Addressed in e273a996854cb3256cfeed832c3f16ff6fd0f534, thank you for the feedback!

ritter-x2a commented on clang/test/SemaCUDA/target-overloads-in-function-prototypes.cu:
```diff
@@ -0,0 +1,703 @@
+// Default arguments for template parameters occur before the target attribute,
+// so we can't identify the "right" overload for them.
+template <typename T = targetdep_t>
+__attribute__((device)) // expected-warning {{target attribute has been ignored for overload resolution}}
```
A point worth noting is this case, where moving the target attribute before the overloaded call (here wrapped in `targetdep_t`) is not possible: in default arguments for template parameters. The rocPRIM library contains this kind of code (and therefore also the introduced warning). I'd be happy about suggestions to improve this.

ritter-x2a replied:
We could mitigate this issue at least for non-type template parameters with default arguments by supporting this syntax to control overload resolution:
```CUDA
template <__attribute__((device)) int V = device_specific_init()>
__attribute__((device)) void foo(void) {
  // ...
}
```
Somewhat surprisingly (to me), trunk already accepts device attributes in non-type template arguments without diagnostic (but seems to ignore them), so this would not introduce new syntax.
So far, I have not encountered type template parameters with target-dependent default arguments -- where this syntax is not possible -- outside of this test. The problematic cases in rocPRIM are non-type template parameters (e.g. the device function `device_warp_size` in the default argument [here](https://github.com/ROCm/rocPRIM/blob/100b3ea65b61edf5af0b7f3fb98b8670cce51c2a/rocprim/include/rocprim/block/block_load_func.hpp#L372C29-L372C45)).

@yxsamliu, do you think that this is a viable way to avoid the warnings, e.g., in rocPRIM?

ritter-x2a replied:
@yxsamliu, I opened #109663 with an alternative solution for the root problem that I'm actually trying to solve (diagnosing wrong uses of `__AMDGCN_WAVEFRONT_SIZE`), which would make PRs #91478, #93546, and #103031 (this one) unnecessary.
Please let me know if you think we should rather pursue #109663.

ritter-x2a commented:
Ping.

ritter-x2a commented:
Ping.

ritter-x2a commented:
Closing this PR in favor of a more comprehensive treatment of the AMDGCN_WAVEFRONT_SIZE situation.
